### ISSUE TITLE
Ability to update or poll for time bounds manually. 

### ISSUE THREAD

**Issue by trusktr** [nasa/openmct#4362]
**Is your feature request related to a problem? Please describe.**

Yes: when making a custom view with a time conductor (scrubber), I want to update the view continually while the timeline is dragged. However it seems that openmct fires the `bounds` event only once I let go of the scrubber. I'd like to get bounds continually while moving my pointer.

**Describe the solution you'd like**

At the moment calling the `time.bounds()` method returns the same bounds at any point while dragging. This is because I do not have a ticking clock, otherwise a ticking clock would cause the time bounds to be up to date.

This ultimately stems from the fact that bounds updates a push based, and require a ticking clock. It would be great if bounds updates were also poll-based.

As an example, here is [Three.js Clock](https://github.com/mrdoob/three.js/blob/dev/src/core/Clock.js).

Ideally, one would be able to call `getTime()`, or the existing `bounds()`, or other similar APIs, and they would simply return the current values.

With an API like that, I could then make my own loop like the following for example:

```js
timeConductorSVG.addEventListener('pointerdown', event => {
  const id = event.pointerId
  let keepGoing = true

  const loop = () => {
    const bounds = this.openmct.time.getBounds()
    
    // ...do something with current bounds...

    if (keepGoing) requestAnimationFrame(loop)
  }

  requestAnimationFrame(loop)

  const onUp = (event) => {
    if (event.pointerId !== id) return // react to the same finger, not other fingers
    keepGoing = false
    document.removeEventListener('pointerup', onUp)
  }

  document.addEventListener('pointerup', onUp)
})
```

In this way I have made my own loop that openmct doesn't have to know about, and I can get current values any time I ask for them.

This technique can be advantageous over clock ticks using `setTimeout` because updates are now in sync with browser rendering.

**Describe alternatives you've considered**

I don't know of any, because after looking around in the code base, I can't see any clean way to get continual bounds updates. Every approach that I know of so far involves re-implementing internal logic (f.e. ticking `DefaultClock` with manual time values but I don't know what time values to pass in, and I would have to implement logic to figure that out based on current conductor zoom level and current bounds, etc)

**Comment by akhenry** [comment]
So the reason you can't poll this in real time is that when you scrub the time conductor, the bounds don't actually update in the Time API until you release the mouse. The Time Conductor _**UI**_ tracks the start and end time in real-time as you scrub, but it doesn't fire a Time API bounds change until you release the mouse. This was by design. We only have one mechanism for pushing out bounds changes (`openmct.time.on('bounds')`) and if we pushed bounds changes out as you scrub it would quickly overwhelm a lot of our views, most of which trigger server queries on bounds changes.

There is actually an `isTick` flag which is true when the bounds change is due to a ticking clock. This flag is used by many of our views to "ignore" the bounds change and not fetch new data from the server. We could potentially re-purpose this flag, or re-cast as something more general that basically says "the bounds are changing but there's no need to fetch data...yet". And then issue another bounds update at the end which would would trigger a historical re-query. This would still require refactoring of our views though, some of them have built in assumptions that these "tick" events only progress time forward.

**Comment by trusktr** [comment]
This is a tricky problem.

> if we pushed bounds changes out as you scrub it would quickly overwhelm a lot of our views, most of which trigger server queries on bounds changes.

This is true. I believe that a query API, for certain cases, would alleviate this, by allowing select UI pieces to query when needed (rather than all UI pieces be force pushed to).

As an example, in Three.js this is how one would use a [`Clock`](https://github.com/mrdoob/three.js/blob/dev/src/core/Clock.js) instance:

```js
requestAnimationFrame(function loop() {
  const time = clock.getElapsedTime()
  const dt = clock.getDelta()
  // ...
  requestAnimationFrame(loop)
})
```

and in this way, any individual component decides if it wants to query. That `Clock` class has no running process: if it is not being queried by a component, it is using zero CPU.

But! Three.js Clock is for a much simpler purpose: only for realtime time, not for scrubbing, and not for historical time.

With only push-based ticking, if the tick rate were to be sped up, then like you suggested all components would need to be modified to be throttled, which wouldn't be ideal.

So what I'm imagining is this:

- We can refactor the clock to make it a queryable clock.
- But this clock also needs to be settable so it can be controlled for scrubbing or historical playback.
- To keep backwards compatibility with existing components, we can make a separate process that ticks itself, queries the clock, then passes the value along like before. In this new setup, the clock itself has no tick, only a time value that is calculated upon query.
- New components, or any refactored components, can begin to rely on the querying approach as needed: for example, a piece of UI can query the clock during drag events, or some component that performs an animation can query a clock only during the animation.

This updated clock system would, at a high level, be a hybrid of the current approach plus the new approach.

Despite all that, I think this is probably lower priority for the time being, but will be useful for 60fps animation when needed.

**Comment by akhenry** [comment]
> Ideally, one would be able to call getTime(), or the existing bounds(), or other similar APIs, and they would simply return the current values.

To summarize for those who didn't read all of the comments above, `openmct.time.bounds()` _does_ return the current value. As does `openmct.time.clock().currentValue()` for a ticking clock (but this is redundant, the bounds are kept up to date). So the API already supports what is being proposed here. 

The issue here is one of UI optimization, scrubbing the time conductor does not update the API until the user has finished scrubbing, for performance reasons.
